### PR TITLE
Meat

### LINKED ISSUE #7: Create Meat
# Endpoint
POST(/meat)

** `userId` is get from auth middleware

## Request body
```
{
    "title": "title",
    "description": "description",
    "maxParticipant": 1,
    "startTime": "2021-01-02T05:40:46.282Z",
    "endTime": "2021-01-02T05:40:46.282Z",
    "preferenceIds": [1,2],
    "locationDTO": {
        "latitude": 51.5074,
        "longitude": 0.1278
    },
    "base64String": "data:image/jpeg;base64,base64ContentHere"
}
```
# Validation
### Form Validation
1. title is not empty, max character is 50
2. description can be null
3. maxParticipant is not negative number (can be null means unlimited)
4. startTime is not empty, startTime is later than now
5. endTime is not empty, endTime is later than startTime
6. location is not empty
7. image is not empty, is image type (not other format e.g. pdf)
8. preferenceId must have at least 1. maximum is 3.

# Business Logic
1. upload image to firebase storage
2. save image to MySQL `storage`
3. save meat into MySQL `meat`
4. save preference to MySQL `meat_preference`
5. save meat location to firestore
6. save the userId in the token as organiser into MySQL `meat_user`

# Response
1. 200: Meat with id 35 saved successfully
2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #8: Update Meat
# Endpoint
PUT(/meat)

## Request body
```
{
    "id": 35,
    "title": "title2",
    "description": "description2",
    "maxParticipant": 2,
    "startTime": "2022-01-02T05:40:46.282Z",
    "endTime": "2022-01-02T05:40:46.282Z",
    "preferenceIds": [3,4],
    "locationDTO": {
        "latitude": 52.5074,
        "longitude": 42.1278
    }
    "base64String": "data:image/jpeg;base64,base64ContentHere"
}
```
### Notes
`locationDTO` and `base64String` is optional in request body.
`preferenceId` is not optional

# Validation
### Show Banner message and hide form if
1. the meat is ended
2. the meat is cancelled
3. the user is not the organiser of meat

### Form Validation
1. title is not empty, max character is 50
2. description can be null
3. maxParticipant is not negative number (can be null means unlimited)
4. startTime is not empty, startTime is later than now, if meat is ongoing/ended cannot change startTime
5. endTime is not empty, endTime is later than startTime
6. location is can be null (means location remain the same)
7. image can be null (means image remain the same), is image type (not other format e.g. pdf)
8. preferenceId must have at least 1. maximum is 3.

# Business Logic
1. if base64String exists in request body, upload image to firebase storage
2. update meat and save to MySQL `meat`
     1. only can update imageStorageId, title, description, maxParticipant, startTime, endTime
     2. auto change lastModifiedDate
3. do meat_preference update business logic
     1. delete all old meat_preference by meatId
     2. save the new meat_preference that is in request body
4. if locationDTO exists in request body, update meat location to firestore

# Response
1. 200: mysqlResponse
2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #9: Cancel Meat
# Endpoint
PUT(/meat/:meatId/cancel)

# Validation
### Only show cancel button if
1. the user is the organizer
2. the meat is not ended (already started still can cancel meat)
3. meat is not canceled

# Business Logic
1. update meat and save to MySQL `meat`
     1. change status to `cancelled`
     2. auto change lastModifiedDate

# Response
1. 200: mysqlResponse
2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #10: Find One Meat
# Endpoint
GET(/meat/:meatId)

** `userId` is get from auth middleware

# Business Logic
1. findOneMeatByMeatId (get meat information)
2. findOneStorageByStorageId (get imageUrl)
3. findOneMeatLocationByMeatId (locationDTO)
4. findAllMeatPreferenceByMeatId
5. mapPreferenceInformation
6. getMeatAnalyticsService by meatId and userId (get totalParticipants, role[organiser, participant, null - haven't join], status[going, not_coming, null - haven't join])

# Response
1. 200:
```
{
    "id": 35,
    "imageUrl": "https://storage.googleapis.com/foodie-vlog-27dc6.appspot.com/meat_images/meat_images_2021-01-07T04%3A54%3A07.565Z.jpeg?GoogleAccessId=firebase-adminsdk-svj0s%40foodie-vlog-27dc6.iam.gserviceaccount.com&Expires=16446988800&Signature=ONOCvPyypc2gYvey3EpjyREObFiuWVe5RXUNtlCsiFj0lQv4yZCyrb1TuEQ1h1VqX14%2BW5I7w7Ki8G88cfuCbLZZ6gek%2BxJRD9YrW02oOjwX3n7QxvPCH9ollL4dG5UKumU2f9dPvsWfH1GrlvWp3n5GKvwPC86JorB%2FGn5%2B%2BFoplzQk9cCxzjTUvmInapV%2BooNXF4OtDcv73ioXzGwlqEQnT0byT1EuxoFgW15u%2F69drzPbyZ0YQYkweHnhAfaBYpZrAOaC2qIocygsuQWN3%2FIbhQS%2FsJ4tCNzg1nt4zp6RdYhWy15615GX8%2BTSTwBM7cBrPevAH2wDDe%2BhkE0SEg%3D%3D",
    "title": "title2",
    "description": "description2",
    "maxParticipant": 2,
    "startTime": "2022-01-02T05:40:46.000Z",
    "endTime": "2022-01-02T05:40:46.000Z",
    "status": "ongoing",
    "preferences": [
          {
               "id": 1,
               "name": "spicy",
               "description": "as hot as you",
          }
     ],
    "locationDTO": {
        "latitude": 52.5074,
        "longitude": 42.1278
    },
    "totalParticipants": 0,
    "role": null,
    "status": null,
    "createdDate": "2021-01-07T04:54:08.000Z",
    "lastModifiedDate": "2021-01-07T04:57:56.000Z"
}
```
2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #11: View my upcoming meat
# Endpoint
GET(/meat)

** `userId` is get from auth middleware

# Business Logic
1. findAllMeatUsersByUserIdAndStatusIsGoing
2. filterMeatIsOngoing (not cancelled)
3. filter meat that is not ended
4. mapMeatDetails
5. mapImageURL

** location is not needed 
** preference is not needed
** meat analytics is not needed (totalParticipant, role)

# Response
1. 200:
```
[{
    "id": 35,
    "imageUrl": "https://storage.googleapis.com/foodie-vlog-27dc6.appspot.com/meat_images/meat_images_2021-01-07T04%3A54%3A07.565Z.jpeg?GoogleAccessId=firebase-adminsdk-svj0s%40foodie-vlog-27dc6.iam.gserviceaccount.com&Expires=16446988800&Signature=ONOCvPyypc2gYvey3EpjyREObFiuWVe5RXUNtlCsiFj0lQv4yZCyrb1TuEQ1h1VqX14%2BW5I7w7Ki8G88cfuCbLZZ6gek%2BxJRD9YrW02oOjwX3n7QxvPCH9ollL4dG5UKumU2f9dPvsWfH1GrlvWp3n5GKvwPC86JorB%2FGn5%2B%2BFoplzQk9cCxzjTUvmInapV%2BooNXF4OtDcv73ioXzGwlqEQnT0byT1EuxoFgW15u%2F69drzPbyZ0YQYkweHnhAfaBYpZrAOaC2qIocygsuQWN3%2FIbhQS%2FsJ4tCNzg1nt4zp6RdYhWy15615GX8%2BTSTwBM7cBrPevAH2wDDe%2BhkE0SEg%3D%3D",
    "title": "title2",
    "description": "description2",
    "maxParticipant": 2,
    "startTime": "2022-01-02T05:40:46.000Z",
    "endTime": "2022-01-02T05:40:46.000Z",
    "status": "ongoing",
    "createdDate": "2021-01-07T04:54:08.000Z",
    "lastModifiedDate": "2021-01-07T04:57:56.000Z"
}]
```
2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #12: Explore meat
# Endpoint
GET(/meat/explore?perferenceId=:preferenceId, latitude=:latitude, longtitude=:longitude)

** `userId` is get from auth middleware

## Query params
1. userId is mandatory
    1. to know default location and default preferences to search for nearby
2. preferenceId is optional, can put if don't want to use default preference of user
3. latitude and longitude is optional, can put if don't want to use default location of user

# Business Logic
1. search nearby 10km using firestore + geofire, using location given in the query param, if no location is given, use user default location (will return an array of MeatLocation and distanceInKM)
2. getPreference, if no preference is found in the query param, go to user profile to find user default preference
3. filter the nearby result, only includes matched preference
4. mapMeatDetails
5. filter to only includes `ongoing` meat
6. filter to only includes meat that has not ended (already started will still be included)
7. filter to only includes meat that still have vacant `total_participant < max_participant`
8. mapImageURL

# Response
1. 200:
```
[] // no nearby meat found
```
```
[{
    "id": 35,
    "imageUrl": "https://storage.googleapis.com/foodie-vlog-27dc6.appspot.com/meat_images/meat_images_2021-01-07T04%3A54%3A07.565Z.jpeg?GoogleAccessId=firebase-adminsdk-svj0s%40foodie-vlog-27dc6.iam.gserviceaccount.com&Expires=16446988800&Signature=ONOCvPyypc2gYvey3EpjyREObFiuWVe5RXUNtlCsiFj0lQv4yZCyrb1TuEQ1h1VqX14%2BW5I7w7Ki8G88cfuCbLZZ6gek%2BxJRD9YrW02oOjwX3n7QxvPCH9ollL4dG5UKumU2f9dPvsWfH1GrlvWp3n5GKvwPC86JorB%2FGn5%2B%2BFoplzQk9cCxzjTUvmInapV%2BooNXF4OtDcv73ioXzGwlqEQnT0byT1EuxoFgW15u%2F69drzPbyZ0YQYkweHnhAfaBYpZrAOaC2qIocygsuQWN3%2FIbhQS%2FsJ4tCNzg1nt4zp6RdYhWy15615GX8%2BTSTwBM7cBrPevAH2wDDe%2BhkE0SEg%3D%3D",
    "title": "title2",
    "description": "description2",
    "maxParticipant": 2,
    "startTime": "2022-01-02T05:40:46.000Z",
    "endTime": "2022-01-02T05:40:46.000Z",
    "status": "ongoing",
    "distance": 5, // 5km away
    "createdDate": "2021-01-07T04:54:08.000Z",
    "lastModifiedDate": "2021-01-07T04:57:56.000Z"
}]
```
** preference is not needed (because if the user has preference filter on, it will be shown ady)
** location is not needed (no need to show latitude and longitude, only need to know how many km away)
** meat analytics is not needed (no need totalParticipants, role)

2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #13: Join Meat
# Endpoint
POST(/meat/:meatId/join)

** `userId` is get from auth middleware

# Validation
### Only show join button if 
1. meat is not ended
    1. _user can still join started meat_
2. user not yet join this meat
3. if `total_participant < max_participant`
4. meat is not cancelled

# Business Logic
1. save the userId in the token as participant into MySQL `meat_user`

# Response
1. 200: MeatUser with id 35 saved successfully
2. 400: BadRequest
3. 500: Unknown Error

### LINKED ISSUE #14: Unjoin Meat
# Endpoint
POST(/meat/:meatId/unjoin)

** `userId` is get from auth middleware

# Validation
### Only show unjoin button if 
1. the user has joined the meat
2. meat is not ended
    1. _user can still unjoin started meat_
3. meat is not canceled

# Business Logic
1. change the status from "going" to "not_coming" from MySQL  `meat_user` where meatId and userId

# Response
1. 200: MySQLserverResponse
2. 400: BadRequest
3. 500: Unknown Error

### PR DESCRIPTION
**Pull request by kachun333**
resolve #7 
resolve #8 
resolve #9 
resolve #10 
resolve #11 
resolve #12 
resolve #13 
resolve #14

### CODE REVIEW

File: src/repository/meatUsers.repostitory.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
+const { connection } = require("../config/mysql");
+const {
+  createPlaceholderString,
+  toMysqlTimestampString,
+} = require("../utils/mysql.utils");
+const MeatUser = require("../entities/meatUser.entity");
+const MeatUserStatus = require("../enums/meatUserStatus.enum");
+const MeatUserRole = require("../enums/meatUserRole.enum");
+
```
Comment (junzai97) [inline src/repository/meatUsers.repostitory.js:9]: Perhaps you have missed out hasMissingKey check.

File: src/services/nearby.service.js [hunk 2]
```diff
@@ -6,24 +6,28 @@ const LocationDTO = require("../dtos/locationDTO.dto");
...
-      doc.latitude,
-      doc.longitude,
-      doc.geohash
+  return results.map(({ doc, distanceInKm }) => {
+    const docData = doc.data();
+    const meatLocation = new MeatLocation(
+      docData.meatId,
+      docData.latitude,
+      docData.longitude,
+      docData.geohash
```
Comment (junzai97) [inline src/services/nearby.service.js:20]: It doesn't seem like we need to create the meatLocation instance as we only require the meatId to be return

File: src/services/userPreference.service.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const {
```
Comment (junzai97) [inline src/services/userPreference.service.js]: Does this class really require? Caused the only method is calling another method from repository.

File: src/repository/meatUsers.repostitory.js (hunk 1 above)
Comment (kachun333) [reply src/repository/meatUsers.repostitory.js:9]: i just pass the id inside tho. 
I didn't pass in an object and save the entire object

File: src/services/meatUser.service.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  findAllMeatUserByUserIdAndStatus,
+} = require("../repository/meatUsers.repostitory");
+const MeatUserRole = require("../enums/meatUserRole.enum");
+const MeatUserStatus = require("../enums/meatUserStatus.enum");
+
+async function createMeatOrganiserService(meatId, userId) {
+  return await createMeatUser(meatId, userId, MeatUserRole.ORGANISER);
+}
+
+async function createMeatParticipantService(meatId, userId) {
```
Comment (junzai97) [inline src/services/meatUser.service.js:14]: No checking for whether user already joined the meat.

File: ddl_script/08022021_meat_preference.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS meat_preference (
+    id BIGINT AUTO_INCREMENT,
+    meat_id BIGINT NOT NULL,
+    preference_id BIGINT NOT NULL,
```
Comment (junzai97) [inline ddl_script/08022021_meat_preference.sql:4]: Should register as Foreign key

**Review by junzai97 — APPROVED** [review]

**Pull request merged**